### PR TITLE
Datasource: Propagate datasource secret decryption errors to the frontend

### LINKED ISSUE #51597: Show detailed error on dashboard if the plugin fails to retrieve a datasource secret
Expand on work done in https://github.com/grafana/grafana/pull/50542 to show detailed errors if secret retrieval fails on the plugin side.

### PR DESCRIPTION
**Pull request by mmandrus**
**What this PR does / why we need it**:
This PR propagates errors caused by secret decryption to the frontend so that we can show useful error messages when there is a secrets plugin problem.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/51597

**Special notes for your reviewer**:

In the current baseline, when decryption errors occur, we continue to try to query the database anyway, and may be successful because of datasource connection caching. This makes it difficult to deliver useful errors to the frontend, and begs the question of whether we _should_ be able to query a datasource if we don't have up-to-date decrypted passwords.

After chatting with @ryantxu, we determined that updating the /ds/query endpoint to fail if secret decryption fails would make the most sense, unless there is a specific reason why we should continue through the code path anyway. 

Will create a unit test once this code solution is approved. 

Corresponds with enterprise PR https://github.com/grafana/grafana-enterprise/pull/3574

### CODE REVIEW

File: pkg/api/datasources.go [hunk 1]
```diff
@@ -626,13 +626,13 @@ func (hs *HTTPServer) checkDatasourceHealth(c *models.ReqContext, ds *datasource
...
 }
 
-func (hs *HTTPServer) decryptSecureJsonDataFn(ctx context.Context) func(ds *datasources.DataSource) map[string]string {
-	return func(ds *datasources.DataSource) map[string]string {
+func (hs *HTTPServer) decryptSecureJsonDataFn(ctx context.Context) func(ds *datasources.DataSource) (map[string]string, error) {
+	return func(ds *datasources.DataSource) (map[string]string, error) {
 		decryptedJsonData, err := hs.DataSourcesService.DecryptedValues(ctx, ds)
 		if err != nil {
 			hs.log.Error("Failed to decrypt secure json data", "error", err)
 		}
```
Comment (papagian) [inline pkg/api/datasources.go:634]: nit
given that the API responses (with status code >= 500) [are logged anyway](https://github.com/grafana/grafana/blob/ef77c93934b5927a29a94469440f4d2ddaa30656/pkg/middleware/logger.go#L69) I don't think that this error logging is no more needed.

File: pkg/api/metrics.go [hunk 2]
```diff
@@ -24,7 +25,13 @@ func (hs *HTTPServer) handleQueryMetricsError(err error) *response.NormalRespons
 	if errors.Is(err, datasources.ErrDataSourceNotFound) {
 		return response.Error(http.StatusNotFound, "Data source not found", err)
 	}
-	var badQuery *query.ErrBadQuery
+
+	var secretsPlugin datasources.ErrDatasourceSecretsPluginUserFriendly
+	if errors.As(err, &secretsPlugin) {
+		return response.Error(http.StatusInternalServerError, fmt.Sprint("Secrets Plugin error: ", err.Error()), err)
+	}
```
Comment (papagian) [inline pkg/api/metrics.go:32]: given that here we only need to check is the error matches the `datasources.ErrDatasourceSecretsPluginUserFriendly`  I think that we need to use `errors.Is()` instead.

**Review by papagian — COMMENTED** [review]
It makes also more sense to me to error in this case but I don't know why we didn't before.
I left some minor comments.

**Review by leandro-deveikis — APPROVED** [review]
LGTM!

**Pull request merged**